### PR TITLE
fix(@angular-devkit/build-angular): update karma builder to use non-deprecated API

### PR DESCRIPTION
**Pull request by alan-agius4**
With this change we update the Karma builder to use the new Server API.

Closes: #20479

### CODE REVIEW

File: packages/angular_devkit/build_angular/src/karma/index.ts [hunk 1]
```diff
@@ -72,105 +72,114 @@ export function execute(
...
+        },
+      );
+      // karma typings incorrectly define start's return value as void
+      // tslint:disable-next-line:no-use-of-empty-return-value
+      const karmaStart = (karmaServer.start() as unknown) as Promise<void>;
+
+      // Cleanup, signal Karma to exit.
+      return () => {
+        // Karma only has the `stop` method start with 3.1.1, so we must defensively check.
+        const karmaServerWithStop = (karmaServer as unknown) as { stop: () => Promise<void> };
```
Comment (devoto13) [inline packages/angular_devkit/build_angular/src/karma/index.ts:177]: nit: I guess this is no longer relevant. Maybe fix in the follow-up PR?

File: packages/angular_devkit/build_angular/src/karma/index.ts [hunk 2]
```diff
@@ -72,105 +72,114 @@ export function execute(
...
+        });
+      }
+
+      karmaOptions.buildWebpack = {
+        options,
+        webpackConfig,
+        // Workaround for https://github.com/karma-runner/karma/issues/3154
+        // When this workaround is removed, user projects need to be updated to use a Karma
+        // version that has a fix for this issue.
+        toJSON: () => { },
```
Comment (devoto13) [inline packages/angular_devkit/build_angular/src/karma/index.ts:138]: nit: The issue has been fixed a while ago, so maybe cleanup in the follow-up PR?

File: packages/angular_devkit/build_angular/src/karma/index.ts [hunk 3]
```diff
@@ -72,105 +72,114 @@ export function execute(
...
+        logger: context.logger,
+      };
+
+      // @types/karma don't include the last parameter.
```
Comment (devoto13) [inline packages/angular_devkit/build_angular/src/karma/index.ts:142]: FYI I'm working on the fix for this and other issues in @types/karma. Should have a PR up soon.

Comment (devoto13) [reply packages/angular_devkit/build_angular/src/karma/index.ts:142]: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52286

File: packages/angular_devkit/build_angular/src/karma/index.ts (hunk 2 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/karma/index.ts:138]: Will do

**Review by clydin — APPROVED** [review]

**Comment by erikhorlings** [comment]
Hey all,
Just wanted to say thanks for the work you put in.

**Comment by erikhorlings** [comment]
Is this change limited to the newest v12, or can it also be merged to the 11.2.x branch?

**Comment by alan-agius4** [comment]
@erikhorlings, this change is only merged in version 12.

**Pull request merged**